Development/nmos/control_protocol_utils.cpp

```diff
@@ -0,0 +1,648 @@
+                    {
+                        // get resource based on the oid
+                        const auto& oid = nmos::fields::nc::oid(member);
+                        const auto& found = find_resource(resources, utility::s2us(std::to_string(oid)));
+                        if (resources.end() != found)
+                        {
+                            find_members_by_role(resources, *found, role, match_whole_string, case_sensitive, recurse, descriptors);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // find members with given class id
+    void find_members_by_class_id(const resources& resources, const nmos::resource& resource, const nc_class_id& class_id_, bool include_derived, bool recurse, web::json::array& descriptors)
+    {
+        auto find_members_by_matching_class_id = [&](const web::json::array& members)
+        {
+            using web::json::value;
+
+            auto match = [&](const web::json::value& descriptor)
+            {
+                const auto& class_id = nmos::details::parse_nc_class_id(nmos::fields::nc::class_id(descriptor));
+
+                if (include_derived) { return !boost::find_first(class_id, class_id_).empty(); }
+                else { return class_id == class_id_; }
+            };
+
+            return boost::make_iterator_range(boost::make_filter_iterator(match, members.begin(), members.end()), boost::make_filter_iterator(match, members.end(), members.end()));
+        };
+
+        if (resource.data.has_field(nmos::fields::nc::members))
+        {
+            auto& members = nmos::fields::nc::members(resource.data);
+
+            auto members_found = find_members_by_matching_class_id(members);
+            for (const auto& member : members_found)
+            {
+                web::json::push_back(descriptors, member);
+            }
+
+            if (recurse)
+            {
+                // do class_id match on all NcBlock(s)
+                for (const auto& member : members)
+                {
+                    if (is_nc_block(nmos::details::parse_nc_class_id(nmos::fields::nc::class_id(member))))
+                    {
+                        // get resource based on the oid
+                        const auto& oid = nmos::fields::nc::oid(member);
+                        const auto& found = find_resource(resources, utility::s2us(std::to_string(oid)));
+                        if (resources.end() != found)
+                        {
+                            find_members_by_class_id(resources, *found, class_id_, include_derived, recurse, descriptors);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // push a control protocol resource into other control protocol NcBlock resource
+    void push_back(control_protocol_resource& nc_block_resource, const control_protocol_resource& resource)
+    {
+        using web::json::value;
+
```

Comment:
Should there be  a write lock?

Reply:
This function is only used by the `node_implementation_init`, which has already had the write lock protecting it.

Will add a note in the function to remind the user to set the write lock in the outer function.